apps/comps/rpc/middleware/cache.ts

```diff
@@ -0,0 +1,70 @@
+import { StandardRPCJsonSerializer } from "@orpc/client/standard";
+import {
+  Context,
+  ErrorMap,
+  Meta,
+  Middleware,
+  ORPCErrorConstructorMap,
+  os,
+} from "@orpc/server";
+import { unstable_cache } from "next/cache";
+import type { Logger } from "pino";
+
+export function cacheMiddleware<
+  TInContext extends Context & { logger: Logger },
+  TInput,
+  TOutput,
+  TErrorMap extends ErrorMap,
+  TMeta extends Meta,
+>(options?: {
+  key?: string[];
+  revalidate?: number;
+  tags?: string[];
+  includeContext?: (context: TInContext) => Record<string, any>;
+}): Middleware<
+  TInContext,
+  Record<never, never>,
+  TInput,
+  TOutput,
+  ORPCErrorConstructorMap<TErrorMap>,
+  TMeta
+> {
+  return os
+    .$context<{ logger: Logger }>()
+    .middleware(async ({ context, next, path }, input) => {
+      const contextValues = options?.includeContext?.(context as TInContext);
+
+      const serializer = new StandardRPCJsonSerializer();
+      const [serialized] = serializer.serialize({
+        input,
+        contextValues,
+      });
```

Comment:
Our input and context may contain non json-compatible properties like `Date` and `bigint`, so we use oRPC's built in serializer to serialize them into json.